src/main/java/teammates/ui/webapi/UpdateFeedbackSessionLogsAction.java

```diff
@@ -34,30 +33,30 @@ public JsonResult execute() {
         List<FeedbackSessionLogEntry> logEntries = logsProcessor.getOrderedFeedbackSessionLogs(null, null,
                 startTime.toEpochMilli(), endTime.toEpochMilli(), null);
 
-        Map<String, Map<String, Map<String, Long>>> lastSavedTimestamps = new HashMap<>();
+        Map<String, Map<String, Map<String, Map<String, Long>>>> lastSavedTimestamps = new HashMap<>();
         for (FeedbackSessionLogEntry logEntry : logEntries) {
+            String courseId = logEntry.getCourseId();
             String email = logEntry.getStudentEmail();
             String fbSessionName = logEntry.getFeedbackSessionName();
             String type = logEntry.getFeedbackSessionLogType();
             Long timestamp = logEntry.getTimestamp();
 
             lastSavedTimestamps.putIfAbsent(email, new HashMap<>());
-            lastSavedTimestamps.get(email).putIfAbsent(fbSessionName, new HashMap<>());
-            Long lastSaved = lastSavedTimestamps.get(email).get(fbSessionName).getOrDefault(type, 0L);
+            lastSavedTimestamps.get(email).putIfAbsent(courseId, new HashMap<>());
+            lastSavedTimestamps.get(email).get(courseId).putIfAbsent(fbSessionName, new HashMap<>());
+            Long lastSaved = lastSavedTimestamps.get(email).get(courseId).get(fbSessionName).getOrDefault(type, 0L);
 
             if (Math.abs(timestamp - lastSaved) > SPAM_FILTER) {
-                lastSavedTimestamps.get(email).get(fbSessionName).put(type, timestamp);
-                FeedbackSessionLog fslEntity = new FeedbackSessionLog(email, fbSessionName,
+                lastSavedTimestamps.get(email).get(courseId).get(fbSessionName).put(type, timestamp);
+                Student student = sqlLogic.getStudentForEmail(courseId, email);
+                FeedbackSession feedbackSession = sqlLogic.getFeedbackSession(fbSessionName, courseId);
```

Comment:
Since it's an optimization problem, it's fine to leave a comment and improve on it later

Reply:
Good catch! Since there is a high probability that the logs belong to the same student/session, I have updated the action to store queried students and feedback sessions in a map.

Reply:
Since we have changed the logs to store the ids, I changed it to use [`getReference()`](https://thorben-janssen.com/jpa-getreference/#the-getreference-method). 

This should make it such that we incur 0 db reads for this part.